packages/ruleset/src/utils/check-composite-schema-for-constraint.js

```diff
@@ -0,0 +1,65 @@
+const isObject = require('./is-object');
+
+/**
+ * This function will return `true` if all possible variations of a (possibly) composite schema are
+ * required to have a constraint, as checked by a lambda function which checks for the constraint on
+ * a simple (non-composite) schema. It does this by recursively checking if:
+ *
+ * - a schema has the constraint itself, or
+ * - one of the schemas it composes with `allOf` has the constraint, or
+ * - all of the schemas it composes with `oneOf` have the constraint, or
+ * - all of the schemas it composes with `anyOf` have the constraint
+ */
+const checkCompositeSchemaForConstraint = (schema, hasConstraint) => {
```

Comment:
@dpopp07 @padamstx I'm not sure if "composite" and "compose" are good ways to differentiate the nested schemas this relates to vs all "subschemas." Looking back at OO concepts, it looks like "composed" schemas would usually mean the equivalent of schemas referred to in `properties`. Maybe we should refer to schemas in `allOf`, etc as "combined" schemas and refer to the combining schema as a "complex" schema? That's what the Swagger documentation does: https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/

Alternatively, we could refer to "complex" schemas containing "applied" (?) schemas based on the JSON Schema terminology: https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.7.5

Reply:
If possible let's continue using the terms that we might already be familiar with instead of new ones.
I think we have typically used the term "composed schema" (or "composed model") to mean a schema that is defined with an allOf/anyOf/oneOf, so I'm ok with the current name `checkCompositeSchemaForConstraint`.  I didn't get tripped up on the name when reviewing the PR :)